File: my-folder/0295-find-median-from-data-stream/solution.py
class MedianFinder:

    def __init__(self):
        self.maxi=[]
        self.mini=[]
        
    def addNum(self, num: int) -> None:
        if self.mini and num>self.mini[0]:
            heapq.heappush(self.mini,num)
        else:
            heapq.heappush(self.maxi,-1*num)
        if len(self.maxi)>len(self.mini)+1:
            val=heapq.heappop(self.maxi)
            heapq.heappush(self.mini,val*-1)

        elif len(self.mini)>len(self.maxi)+1:
            val=heapq.heappop(self.mini)
            heapq.heappush(self.maxi,-1*val)

    def findMedian(self) -> float:
        if len(self.mini)>len(self.maxi):
            return self.mini[0]
        elif len(self.maxi)>len(self.mini):
            return -1*self.maxi[0]

        else:
            return (self.mini[0]+-1*self.maxi[0])/2.0


# Your MedianFinder object will be instantiated and called as such:
# obj = MedianFinder()
# obj.addNum(num)
# param_2 = obj.findMedian()

File: my-folder/0295-find-median-from-data-stream/solution.java
class MedianFinder {
    PriorityQueue<Integer> maxHeap;
    PriorityQueue<Integer> minHeap;
    boolean evenFlag = true;
    public MedianFinder() {
        maxHeap = new PriorityQueue<>();
        minHeap = new PriorityQueue<>((a, b) -> b - a);
    }
    
    public void addNum(int num) {
        if(evenFlag){
            minHeap.offer(num);
            maxHeap.offer(minHeap.poll());
        }
        else{
            maxHeap.offer(num);
            minHeap.offer(maxHeap.poll());
        }
        evenFlag = !evenFlag;
    }
    
    public double findMedian() {
        if(maxHeap.size() > minHeap.size()){
            return maxHeap.peek();
        }
        
        return (double)(maxHeap.peek() + minHeap.peek()) / 2;       
    }

}

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder obj = new MedianFinder();
 * obj.addNum(num);
 * double param_2 = obj.findMedian();
 */
